services/motion/explore/explore.go

```diff
@@ -0,0 +1,577 @@
+// moveResponse is the message type returned by both channels: obstacle detection and plan execution.
+// It contains a bool stating if an obstacle was found or the execution was complete and/or any
+// associated error.
+type moveResponse struct {
+	err     error
+	success bool
+}
+
+// inputEnabledActuator is an actuator that interacts with the frame system.
+// This allows us to figure out where the actuator currently is and then
+// move it. Input units are always in meters or radians.
+type inputEnabledActuator interface {
+	resource.Actuator
+	referenceframe.InputEnabled
+}
+
+// obstacleDetectorObject provides a map for matching vision services to any and all cameras names they use.
+type obstacleDetectorObject map[vision.Service]resource.Name
+
+// ErrNotImplemented is thrown when an unreleased function is called.
+var ErrNotImplemented = errors.New("function coming soon but not yet implemented")
+
+// Config describes how to configure the service. As the explore motion service is not being surfaced to users, this is blank.
+type Config struct{}
+
+// Validate here adds a dependency on the internal framesystem service.
+func (c *Config) Validate(path string) ([]string, error) {
+	return []string{framesystem.InternalServiceName.String()}, nil
+}
+
+// NewExplore returns a new explore motion service for the given robot.
+func NewExplore(ctx context.Context, deps resource.Dependencies, conf resource.Config, logger logging.Logger) (motion.Service, error) {
+	ms := &explore{
+		Named:                 conf.ResourceName().AsNamed(),
+		logger:                logger,
+		obstacleResponseChan:  make(chan moveResponse),
+		executionResponseChan: make(chan moveResponse),
+		backgroundWorkers:     &sync.WaitGroup{},
+	}
+
+	if err := ms.Reconfigure(ctx, deps, conf); err != nil {
+		return nil, err
+	}
+	return ms, nil
+}
+
+// Reconfigure updates the explore motion service when the config has changed.
+func (ms *explore) Reconfigure(
+	ctx context.Context,
+	deps resource.Dependencies,
+	conf resource.Config,
+) (err error) {
+	ms.mutex.Lock()
+	defer ms.mutex.Unlock()
+	// Iterate over dependencies and store components and services along with the frame service directly
+	components := make(map[resource.Name]resource.Resource)
+	services := make(map[resource.Name]resource.Resource)
+	for name, dep := range deps {
+		switch dep := dep.(type) {
+		case framesystem.Service:
+			ms.fsService = dep
+		case vision.Service:
+			services[name] = dep
+		default:
+			components[name] = dep
+		}
+	}
+
+	ms.components = components
+	ms.services = services
+	return nil
+}
+
+type explore struct {
+	resource.Named
+
+	fsService  framesystem.Service
+	components map[resource.Name]resource.Resource
+	services   map[resource.Name]resource.Resource
+	logger     logging.Logger
+
+	processCancelFunc     context.CancelFunc
+	obstacleResponseChan  chan moveResponse
+	executionResponseChan chan moveResponse
+	backgroundWorkers     *sync.WaitGroup
+
+	mutex       sync.Mutex
```

Comment:
fsService, components, and services should be declared under mutex.

Reply:
I would also add a comment that mutex protects these members and additional ensures that only one call to Move or Reconfigure may execute at a time.

Reply:
I had chosen to not include those as they should remain unaffected by any element other than reconfigure. However, you are correct that they are (and should be) mutex lock protected if another function  chooses to change them